### PR TITLE
Add model field pattern

### PR DESCRIPTION
**Pull request by zimengyang**
* Change all model field pattern, name should only contain undersocre, dash, letter or number and starts with alphabetical characters  
* Translate regex pattern to user-friendly msg 
* Add name validation for eventdriver

```
$ dispatch create secret my.secret ./secret.json
[Code: 605] name should start with letter or number and may only contain letters, numbers, underscores and dashes
```

### CODE REVIEW

File: pkg/api/v1/field_patterns.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+	// regex pattern
+	Pattern string `json:"pattern"`
+
+	// translated msg presented to user
+	Message string `json:"message"`
+}
+
+// FieldPatternWordNumberDashOnly letter, underscore, number and dash
+var FieldPatternWordNumberDashOnly = FieldPattern{
```
Comment (tenczar) [inline pkg/api/v1/field_patterns.go]: I'm not sure this name is entirely helpful. `FieldPatternWordNumberDashOnly` and `FieldPatternLetterNumberDashOnly` differ at `Letter` and `Word, where the actual difference is one accepts underscores and the other doesn't.

File: pkg/api/v1/field_patterns.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// FieldPatternLetterNumberDashOnly letter, number and dash
+var FieldPatternLetterNumberDashOnly = FieldPattern{
+	Pattern: `^[a-zA-Z0-9][a-zA-Z0-9\-]*$`,
+	Message: " should start with letter or number and may only contain letters, numbers and dashes",
+}
+
+// Validate validates field pattern and return error with translated msg
+func (p FieldPattern) Validate(fieldName, field string) error {
+	if err := validate.Pattern(fieldName, "body", field, p.Pattern); err != nil {
+		return errors.New(err.Code(), fieldName+p.Message)
```
Comment (tenczar) [inline pkg/api/v1/field_patterns.go]: won't this make the error message look like, for example, `nameshould start with ...`?

File: pkg/api/v1/field_patterns.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// FieldPatternWordNumberDashOnly letter, underscore, number and dash
+var FieldPatternWordNumberDashOnly = FieldPattern{
+	Pattern: `^[\w\d][\w\d\-]*$`,
+	Message: " should start with letter or number and may only contain letters, numbers, underscores and dashes",
+}
+
+// FieldPatternLetterNumberDashOnly letter, number and dash
+var FieldPatternLetterNumberDashOnly = FieldPattern{
```
Comment (tenczar) [inline pkg/api/v1/field_patterns.go]: Do we ever use this?

Comment (zimengyang) [reply pkg/api/v1/field_patterns.go]: For now, this one was only used in EventDriver name.

File: pkg/api/v1/field_patterns.go (hunk 2 above)
Comment (zimengyang) [reply pkg/api/v1/field_patterns.go]: A space is in the beginning of message. But this is not good, I will change this to Join(). Thanks!

File: pkg/api/v1/field_patterns.go (hunk 1 above)
Comment (zimengyang) [reply pkg/api/v1/field_patterns.go]: I agree this is somehow confusing. The LetterNumberDashOnly one is only used in EventDriver. The rest of model name will use WordNumberDashOnly. What do you think if I change them to `FieldPatternName` and `FieldPatternNameNoUnderscore`?

Comment (tenczar) [reply pkg/api/v1/field_patterns.go]: These seem reasonable.

Comment (zimengyang) [reply pkg/api/v1/field_patterns.go]: Thanks, will do!

**Pull request merged**